core/services/offchainreporting/oracle.go

```diff
@@ -0,0 +1,198 @@
+package offchainreporting
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/jinzhu/gorm"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-peerstore/pstoremem"
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/store/offchainreportingdb"
+	"github.com/smartcontractkit/chainlink/core/store/orm"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	ocrnetworking "github.com/smartcontractkit/offchain-reporting/lib/networking"
+	ocr "github.com/smartcontractkit/offchain-reporting/lib/offchainreporting"
+	ocrtypes "github.com/smartcontractkit/offchain-reporting/lib/offchainreporting/types"
+	ocrcontracts "github.com/smartcontractkit/offchain-reporting/lib/prototype/contracts"
+)
+
+const JobType job.Type = "offchainreporting"
+
+func RegisterJobType(
+	db *gorm.DB,
+	config *orm.Config,
+	keyStore *KeyStore,
+	jobSpawner job.Spawner,
+	pipelineRunner pipeline.Runner,
+	ethClient eth.Client,
+	logBroadcaster eth.LogBroadcaster,
+) {
+	jobSpawner.RegisterDelegate(
+		NewJobSpawnerDelegate(db, config, keyStore, pipelineRunner, ethClient, logBroadcaster),
+	)
+}
+
+type jobSpawnerDelegate struct {
+	db             *gorm.DB
+	config         *orm.Config
+	keyStore       *KeyStore
+	pipelineRunner pipeline.Runner
+	ethClient      eth.Client
+	logBroadcaster eth.LogBroadcaster
+}
+
+func NewJobSpawnerDelegate(
+	db *gorm.DB,
+	config *orm.Config,
+	keyStore *KeyStore,
+	pipelineRunner pipeline.Runner,
+	ethClient eth.Client,
+	logBroadcaster eth.LogBroadcaster,
+) *jobSpawnerDelegate {
+	return &jobSpawnerDelegate{db, config, keyStore, pipelineRunner, ethClient, logBroadcaster}
+}
+
+func (d jobSpawnerDelegate) JobType() job.Type {
+	return JobType
+}
+
+func (d jobSpawnerDelegate) ToDBRow(spec job.Spec) models.JobSpecV2 {
```

Comment:
I dont really understand why we need a ToDBRow and FromDBRow.

In fact, I'm not really sure why job.Spec and models.JobSpecV2 are different things.

Reply:
The design here was due to having to break some import cycles between `models` and `job` iirc.  Wrestled with it for a good while and this was the cleanest way I could come up with.  The issue should be 100% alleviated when we get generics next year.  In the meantime it's definitely become clear to me that any DB-specific POD structs have to live in `models`, and any functionality we want to adorn those structs with should live in their respective packages.  Or we can dump the entire codebase into a single folder, which I wouldn't necessarily be against as long as we name files well

Reply:
In that case how about an interface method `ToModel` and `FromModel`?